### PR TITLE
Issue 704 surface extrapolation

### LINKED ISSUE #704: Make particle surface extrapolation 2nd order
**Summary**
Make particle surface extrapolation 2nd order

 **Motivation**
Increase accuracy of particle surface concentration to help match comsol better

### PR DESCRIPTION
**Pull request by Scottmar93**
# Description

Updated the boundary value surface extrapolation to a quadratic expansion (a mathematica notebook calculating the components can be found at: https://github.com/Scottmar93/extrapolation-coefficents/tree/master). Also extended to be valid for non-uniform grids. 

I haven't implemented the boundary flux equivalent as only need the surface vals to improve comparison with comsol

I have stuck an option within the boundary value symbol  to turn on quadratic or linear extrapolation. Whilst it is more related to the discretization, I think this is the easiest place to put it to access the option.

Fixes #704 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/master/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ flake8`
- [x] All tests pass: `$ python run-tests.py --unit`
- [x] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by Scottmar93** [comment]
I have put the extrapolation forms in and added options for spatial methods. The new extrapolations work on some toy examples in the tests but they don't seem to recover the correct behaviour when put into models.

File: pybamm/discretisations/discretisation.py [hunk 1]
```diff
@@ -7,6 +7,18 @@
 from scipy.sparse import block_diag, csr_matrix
 
 
+def has_bc_condition_of_form(symbol, side, bcs, form):
```
Comment (tinosulzer) [inline pybamm/discretisations/discretisation.py]: if being picky, `bc_condition` should be `bc` or `boundary_condition` ....

File: pybamm/spatial_methods/finite_volume.py [hunk 2]
```diff
@@ -586,29 +590,202 @@ def boundary_value_or_flux(self, symbol, discretised_child):
...
+
+        dxN = edges[-1] - nodes[-1]
+        dxNm1 = submesh_list[0].d_nodes[-1]
+        dxNm2 = submesh_list[0].d_nodes[-2]
+
+        child = symbol.child
+
         # Create submatrix to compute boundary values or fluxes
+        # Derivation of extrapolation formula can be found at:
+        # https://github.com/Scottmar93/extrapolation-coefficents/tree/master
```
Comment (tinosulzer) [inline pybamm/spatial_methods/finite_volume.py]: note mathematica doesn't render on github preview

File: pybamm/spatial_methods/scikit_finite_element.py [hunk 3]
```diff
@@ -322,14 +326,17 @@ def integral_form(v, dv, w):
...
         """
         Returns the average value of the symbol over the negative tab ("negative tab")
         or the positive tab ("positive tab") in the Finite Element Method.
 
         Overwrites the default :meth:`pybamm.SpatialMethod.boundary_value`
         """
 
+        if bcs is None:
+            bcs = {}
+
```
Comment (tinosulzer) [inline pybamm/spatial_methods/scikit_finite_element.py]: why is this needed here?

File: pybamm/spatial_methods/spatial_method.py [hunk 4]
```diff
@@ -291,12 +305,17 @@ def boundary_value_or_flux(self, symbol, discretised_child):
             The boundary value or flux symbol
         discretised_child : :class:`pybamm.StateVector`
             The discretised variable from which to calculate the boundary value
+        bcs : dict
+            The boundary conditions
```
Comment (tinosulzer) [inline pybamm/spatial_methods/spatial_method.py]: maybe add a sentence to say that, if available, these are used to improve the accuracy of the boundary value or flux

File: tests/unit/test_spatial_methods/test_finite_volume/test_extrapolation.py [hunk 5]
```diff
@@ -0,0 +1,518 @@
+#
```
Comment (tinosulzer) [inline tests/unit/test_spatial_methods/test_finite_volume/test_extrapolation.py]: good tests!

File: examples/scripts/compare_extrapolations.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import pybamm
+
+
+x_n = pybamm.standard_spatial_vars.x_n
+x_s = pybamm.standard_spatial_vars.x_s
+x_p = pybamm.standard_spatial_vars.x_p
+r_n = pybamm.standard_spatial_vars.r_n
```
Comment (rtimms) [inline examples/scripts/compare_extrapolations.py]: do you need r for lead-acid model?
